### PR TITLE
Onlycoldstarts

### PR DESCRIPTION
**Pull request by mriccia**

### CODE REVIEW

File: lambda/executor.js [hunk 1]
```diff
@@ -152,11 +169,15 @@ const runInParallel = async({num, lambdaARN, lambdaAlias, payloads, preARN, post
...
 
-const runInSeries = async({num, lambdaARN, lambdaAlias, payloads, preARN, postARN, sleepBetweenRunsMs}) => {
+const runInSeries = async({num, lambdaARN, lambdaAlias, payloads, preARN, postARN, onlyColdStarts, sleepBetweenRunsMs }) => {
     const results = [];
     for (let i = 0; i < num; i++) {
+        let aliasToInvoke = utils.buildAliasString(lambdaAlias, onlyColdStarts, i);
         // run invocations in series
-        const {invocationResults, actualPayload} = await utils.invokeLambdaWithProcessors(lambdaARN, lambdaAlias, payloads[i], preARN, postARN);
+        if (onlyColdStarts){
+            await utils.waitForAliasActive(lambdaARN, aliasToInvoke);
```
Comment (alexcasalboni) [inline lambda/executor.js:174]: no try-catch needed here? if it's not needed, I'd remove it above too :)

File: lambda/initializer.js [hunk 2]
```diff
@@ -8,24 +8,86 @@ const defaultPowerValues = process.env.defaultPowerValues.split(',');
...
+            count: lambdaFunctionsToSet.length,
+            continue: true,
+        },
+        powerValues: powerValues
+    }
+    return returnObj;
+};
 
-    return powerValues;
+module.exports.versionPublisher = async(event, context) => {
```
Comment (alexcasalboni) [inline lambda/initializer.js]: ideally this would be in a new file, to keep the one-function-per-file convention :)

File: template.yml [hunk 3]
```diff
@@ -131,6 +131,26 @@ Resources:
...
+      CodeUri: lambda
+      Handler: initializer.versionPublisher
+      Layers:
+        - !Ref SDKlayer
+      Policies:
+        - AWSLambdaBasicExecutionRole # Only logs
+        - Version: '2012-10-17' # allow Lambda actions
+          Statement:
+            - Effect: Allow
+              Action:
```
Comment (alexcasalboni) [inline template.yml]: Let's check which of these actions are still needed for the Initializer function above.

I think now the initializer is only reading and the publisher is only writing?

File: lambda/initializer.js [hunk 4]
```diff
@@ -8,24 +8,86 @@ const defaultPowerValues = process.env.defaultPowerValues.split(',');
...
 
     // reminder: configuration updates must run sequentially
     // (otherwise you get a ResourceConflictException)
-    for (let value of powerValues){
-        const alias = 'RAM' + value;
-        await utils.createPowerConfiguration(lambdaARN, value, alias);
+    for (let powerValue of powerValues){
+        const baseAlias = 'RAM' + powerValue;
+        if (!onlyColdStarts){
+            lambdaFunctionsToSet.push({lambdaARN: lambdaARN, powerValue: powerValue, envVars: envVars, alias: baseAlias});
```
Comment (alexcasalboni) [inline lambda/initializer.js]: Splitting this logic into two separate function has a few security/privacy implications.

My concern here is that we are not only reading/modifying the environment variables of the given Lambda Function, but also passing them into the state machine input/output (and will likely end up into the overall state machine output too?). This could create complications when it comes to logging potentially sensitive variables/secrets to CloudWatch.

At the same time, I cannot think of a great alternative, besides refactoring this to be within the same Initializer function, with the same iterator logic, but then reading/updating the configuration on the spot.

In other words, instead of reading everything upfront to generate the full `initConfigurations` list, we would still iterate `num` times, and invoke `utils.getLambdaPower` and `utils.createPowerConfiguration` each time. I know it's a waste of time to invoke `utils.getLambdaPower` every time, but I think that's a better problem to have than passing around env variables & secrets.

File: lambda/utils.js [hunk 5]
```diff
@@ -69,14 +77,18 @@ module.exports.verifyAliasExistance = async(lambdaARN, alias) => {
...
+module.exports.createPowerConfiguration = async(lambdaARN, value, alias, envVars) => {
     try {
-        await utils.setLambdaPower(lambdaARN, value);
+        await utils.setLambdaPower(lambdaARN, value, envVars);
 
         // wait for functoin update to complete
         await utils.waitForFunctionUpdate(lambdaARN);
 
         const {Version} = await utils.publishLambdaVersion(lambdaARN);
+        if (typeof alias === 'undefined'){
```
Comment (alexcasalboni) [inline lambda/utils.js:88]: When can this happen?

**Review by alexcasalboni — CHANGES_REQUESTED** [review]
I think the overall approach makes sense, but I'd try to limit the amount of info we need to pass around as input/output while looping (see more detailed comment).

File: lambda/executor.js (hunk 1 above)
Comment (mriccia) [reply lambda/executor.js:174]: The one above was for debug purposes - I will remove both

File: lambda/initializer.js (hunk 2 above)
Comment (mriccia) [reply lambda/initializer.js]: I will fix this

File: lambda/utils.js (hunk 5 above)
Comment (mriccia) [reply lambda/utils.js:88]: This happens when we want to create a lambda version with the original power values (before Power Tuner modifications), and we don't want to setup an Alias for it (e.g. it'll be $LATEST)

Comment (alexcasalboni) [reply lambda/utils.js:88]: ok cool, I'd add a comment about it :) it will be harder to guess/remember in 6 months

**Comment by mriccia** [comment]
Closing this PR, please refer to the PR on the main project repo https://github.com/alexcasalboni/aws-lambda-power-tuning/pull/206
